File: dbflow/src/main/java/com/raizlabs/android/dbflow/config/TableConfig.java
package com.raizlabs.android.dbflow.config;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.raizlabs.android.dbflow.sql.queriable.ListModelLoader;
import com.raizlabs.android.dbflow.sql.queriable.SingleModelLoader;
import com.raizlabs.android.dbflow.sql.saveable.ModelSaver;
import com.raizlabs.android.dbflow.structure.ModelAdapter;

import java.util.List;

/**
 * Description: Represents certain table configuration options. This allows you to easily specify
 * certain configuration options for a table.
 */
public final class TableConfig<TModel> {

    public static <TModel> TableConfig.Builder<TModel> builder(Class<TModel> tableClass) {
        return new TableConfig.Builder<>(tableClass);
    }

    private final Class<TModel> tableClass;
    private final ModelSaver<TModel> modelSaver;
    private final SingleModelLoader<TModel> singleModelLoader;
    private final ListModelLoader<TModel> listModelLoader;

    TableConfig(Builder<TModel> builder) {
        tableClass = builder.tableClass;
        modelSaver = builder.modelAdapterModelSaver;
        singleModelLoader = builder.singleModelLoader;
        listModelLoader = builder.listModelLoader;
    }

    @NonNull
    public Class<?> tableClass() {
        return tableClass;
    }

    @Nullable
    public ModelSaver<TModel> modelSaver() {
        return modelSaver;
    }

    @Nullable
    public ListModelLoader<TModel> listModelLoader() {
        return listModelLoader;
    }

    @Nullable
    public SingleModelLoader<TModel> singleModelLoader() {
        return singleModelLoader;
    }

    public static final class Builder<TModel> {

        final Class<TModel> tableClass;
        ModelSaver<TModel> modelAdapterModelSaver;
        SingleModelLoader<TModel> singleModelLoader;
        ListModelLoader<TModel> listModelLoader;

        public Builder(@NonNull Class<TModel> tableClass) {
            this.tableClass = tableClass;
        }

        /**
         * Define how the {@link ModelAdapter} saves data into the DB from its associated {@link TModel}. This
         * will override the default.
         */
        @NonNull
        public Builder<TModel> modelAdapterModelSaver(@NonNull ModelSaver<TModel> modelSaver) {
            this.modelAdapterModelSaver = modelSaver;
            return this;
        }

        /**
         * Define how the table loads single models. This will override the default.
         */
        @NonNull
        public Builder<TModel> singleModelLoader(@NonNull SingleModelLoader<TModel> singleModelLoader) {
            this.singleModelLoader = singleModelLoader;
            return this;
        }

        /**
         * Define how the table loads a {@link List} of items. This will override the default.
         */
        @NonNull
        public Builder<TModel> listModelLoader(@NonNull ListModelLoader<TModel> listModelLoader) {
            this.listModelLoader = listModelLoader;
            return this;
        }

        /**
         * @return A new {@link TableConfig}. Subsequent calls to this method produce a new instance
         * of {@link TableConfig}.
         */
        @NonNull
        public TableConfig build() {
            return new TableConfig<>(this);
        }
    }
}

File: dbflow/src/main/java/com/raizlabs/android/dbflow/config/TableConfig.kt
package com.raizlabs.android.dbflow.config

import com.raizlabs.android.dbflow.adapter.queriable.ListModelLoader
import com.raizlabs.android.dbflow.adapter.queriable.SingleModelLoader
import com.raizlabs.android.dbflow.adapter.saveable.ModelSaver
import com.raizlabs.android.dbflow.adapter.ModelAdapter

/**
 * Description: Represents certain table configuration options. This allows you to easily specify
 * certain configuration options for a table.
 */
class TableConfig<T : Any>(val tableClass: Class<T>,
                           val modelSaver: ModelSaver<T>? = null,
                           val singleModelLoader: SingleModelLoader<T>? = null,
                           val listModelLoader: ListModelLoader<T>? = null) {

    internal constructor(builder: Builder<T>) : this(
            tableClass = builder.tableClass,
            modelSaver = builder.modelAdapterModelSaver,
            singleModelLoader = builder.singleModelLoader,
            listModelLoader = builder.listModelLoader
    )

    /**
     * Table builder for java consumers. use [TableConfig] directly if calling from Kotlin.
     */
    class Builder<T : Any>(internal val tableClass: Class<T>) {
        internal var modelAdapterModelSaver: ModelSaver<T>? = null
        internal var singleModelLoader: SingleModelLoader<T>? = null
        internal var listModelLoader: ListModelLoader<T>? = null

        /**
         * Define how the [ModelAdapter] saves data into the DB from its associated [T]. This
         * will override the default.
         */
        fun modelAdapterModelSaver(modelSaver: ModelSaver<T>) = apply {
            this.modelAdapterModelSaver = modelSaver
        }

        /**
         * Define how the table loads single models. This will override the default.
         */
        fun singleModelLoader(singleModelLoader: SingleModelLoader<T>) = apply {
            this.singleModelLoader = singleModelLoader
        }

        /**
         * Define how the table loads a [List] of items. This will override the default.
         */
        fun listModelLoader(listModelLoader: ListModelLoader<T>) = apply {
            this.listModelLoader = listModelLoader
        }

        /**
         * @return A new [TableConfig]. Subsequent calls to this method produce a new instance
         * of [TableConfig].
         */
        fun build(): TableConfig<*> = TableConfig(this)
    }

    companion object {

        @JvmStatic
        fun <T : Any> builder(tableClass: Class<T>): TableConfig.Builder<T> =
                TableConfig.Builder(tableClass)
    }
}
